File: weights-api.py
import os
import time
import asyncio
import aiohttp
import json  # Import the json module
from typing import Optional, Callable, Any, Dict

class WeightsApi:
    def __init__(self, api_key: Optional[str] = None, endpoint: Optional[str] = None):
        self.api_key = api_key
        self.endpoint = endpoint or os.getenv('WEIGHTS_UNOFFICIAL_ENDPOINT', 'http://localhost:3000')
        self.session: Optional[aiohttp.ClientSession] = None

    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()

    async def api_call(self, path: str, method: str = 'GET', body: Optional[Dict] = None) -> aiohttp.ClientResponse:
        """
        Makes an async HTTP request to the API endpoint
        """
        if not self.session or self.session.closed:
            self.session = aiohttp.ClientSession()

        headers = {
            'Content-Type': 'application/json',
            'x-api-key': str(self.api_key)
        }
        
        url = self.endpoint + path
        
        try:
            if method == 'GET' and body:
                params = {}
                for key, value in body.items():
                    if value is not None:
                        params[key] = str(value)  # Convert values to string
                async with self.session.get(url, params=params, headers=headers) as response:
                    response.raise_for_status()
                    return response
            else:
                async with self.session.request(method, url, json=body, headers=headers) as response:
                    response.raise_for_status()
                    return response
        except aiohttp.ClientError as e:
            raise Exception(f"Weights API Error: {e}")

    async def _check_health(self):
        """
        Checks the health of the API and raises an exception if it's not reachable.
        """
        try:
            await self.get_health_data()
        except Exception as e:
            raise Exception(f"Weights API Error: The API is not reachable. Please check your connection or the API status: {e}")

    async def get_health_data(self) -> Dict:
        """
        Retrieves health status of the API
        """
        try:
            response = await self.api_call('/health')
            return await response.json()
        except Exception as e:
            raise Exception(f"Weights API Error: {e}")

    async def call_with_health_check(self, api_call: Callable[[], Any]) -> Any:
        """
        Wraps API calls with health check
        """
        try:
            await self._check_health()
            return await api_call()
        except Exception as e:
            raise Exception(f"Weights API Error: The API is not reachable. Please check your connection or the API status: {e}")

    async def get_status(self, image_id: str) -> Dict:
        """
        Gets the status of a specific image
        """
        async def call():
            response = await self.api_call(f'/status/{image_id}')
            return await response.json()
        return await self.call_with_health_check(call)

    async def get_quota(self) -> str:
        """
        Retrieves quota information
        """
        async def call():
            response = await self.api_call('/quota')
            return await response.text()
        return await self.call_with_health_check(call)

    async def search_loras(self, query: str) -> Dict:
        """
        Searches for Lora models
        """
        async def call():
            response = await self.api_call('/search-loras', method='GET', body={'query': query})
            return await response.json()
        return await self.call_with_health_check(call)

    async def generate_image(self, prompt: str, lora_name: Optional[str] = None) -> Dict:
        """
        Generates an image based on parameters
        """
        params = {'prompt': prompt, 'loraName': lora_name}
        async def call():
            response = await self.api_call('/generateImage', method='GET', body=params)
            return await response.json()
        return await self.call_with_health_check(call)

    async def generate_progressive_image(self, prompt: str, lora_name: Optional[str] = None, 
                                      callback: Callable[[str, Dict], None] = lambda status, data: None) -> Dict:
        """
        Generates a progressive image with status updates
        """
        await self._check_health()

        result = await self.generate_image(prompt, lora_name)
        image_id = result['imageId']
        status_response = await self.get_status(image_id)
        status = status_response.get('status')
        callback(status, {'imageId': image_id})

        old_modified_date = None
        while True:
            await asyncio.sleep(0.1)
            status_response = await self.get_status(image_id)
            status = status_response.get('status')
            last_modified_date = status_response.get('lastModifiedDate')
            error = status_response.get('error')
            
            if old_modified_date != last_modified_date:
                old_modified_date = last_modified_date
                callback(status, {'imageId': image_id})
            
            if status == 'COMPLETED':
                break
            
            if status == 'FAILED':
                raise Exception(f"Image generation failed: {error}")
        return status_response

File: libs/weights-api.ts
import fetch from "node-fetch";
import { RequestInit } from "node-fetch";

export enum HttpMethod {
  GET = "GET",
  POST = "POST",
}

export enum ImageStatus {
  PENDING = "PENDING",
  PROCESSING = "PROCESSING",
  COMPLETED = "COMPLETED",
  FAILED = "FAILED",
}

export interface GenerationResult {
  imageId: string;
}

export interface HealthData {
  status: string;
}

export interface Lora {
  name: string;
  description: string;
  tags: string[];
  image: string;
}


export interface AudioModel {
  id: string,
  title: string,
  content: string,
  image: string,
}


export interface Params {
  [key: string]: object | string | null;
}

export interface GenerationParams extends Params {
  prompt: string;
  loraName: string | null;
}

export interface StatusResponse {
  status: ImageStatus;
  lastModifiedDate?: string;
  error?: string;
}

export class WeightsApi {
  constructor(apiKey: string | null, endpoint: string | null = null) {
    this.endpoint = endpoint;
    this.apiKey = apiKey;
  }
  private apiKey: string | null = null;
  private endpoint: string | null = null;
  /**
   * Makes an HTTP request to the API endpoint.
   * @param path - The API endpoint path.
   * @param method - The HTTP method (default: 'GET').
   * @param body - The request body (optional).
   * @returns Promise<Response>
   */
  async apiCall(
    path: string,
    method: HttpMethod = HttpMethod.GET,
    body: Params | null = null,
  ) {
    const options: RequestInit = {
      method,
      headers: {
        "Content-Type": "application/json",
        "x-api-key": `${this.apiKey}`,
      },
    };
    let url = this.endpoint + path;
    if (method === HttpMethod.GET && body) {
      const params = new URLSearchParams();
      for (const key in body) {
        if (Object.prototype.hasOwnProperty.call(body, key)) {
          if (body[key] === null) {
            continue;
          }
          params.append(key, String(body[key]));
        }
      }
      url += "?" + params.toString();
    } else if (body) {
      options.body = JSON.stringify(body);
    }
    const response = await fetch(url, options);
    if (response.ok) {
      return response;
    } else {
      throw new Error(
        `Error: ${response.status} - ${JSON.stringify(response)}`,
      );
    }
  }

  /**
   * Retrieves health status of the API.
   * @returns Promise with health data.
   */
  async getHealthData(): Promise<HealthData> {
    try {
      const response = await this.apiCall("/health", HttpMethod.GET);
      return await response.json();
    } catch (error: Error | unknown) {
      throw new Error(
        `Weights API Error: The API is not reachable. Please check your connection or the API status. ${error}`,
      );
    }
  }

  /**
   * Wraps API calls with health check
   * @param apiCall - The API call to make
   * @returns Promise<T>
   */
  async callWithHealthCheck<T>(apiCall: () => Promise<T>): Promise<T> {
    try {
      await this.getHealthData();
      return await apiCall();
    } catch (error: Error | unknown) {
      if (error instanceof Error) {
        throw new Error(
          `Weights API Error: The API is not reachable. Please check your connection or the API status.`,
        );
      }
      throw error;
    }
  }

  /**
   * Gets the status of a specific image.
   * @param params - Object containing imageId.
   * @returns Promise with status information.
   */
  getStatus = async (params: GenerationResult): Promise<StatusResponse> => {
    return this.callWithHealthCheck(() =>
      this.apiCall("/status/" + params.imageId, HttpMethod.GET).then(
        (response) => response.json(),
      ),
    );
  };

  /**
   * Retrieves quota information.
   * @returns Promise with quota data.
   */
  getQuota = async (): Promise<string> => {
    return this.callWithHealthCheck(() =>
      this.apiCall("/quota", HttpMethod.GET).then((response) =>
        response.text(),
      ),
    );
  };

  /**
   * Searches for Lora models.
   * @param params - Object containing search query.
   * @returns Promise with search results.
   */
  searchLoras = async (params: { query: string }): Promise<Lora[]> => {
    return this.callWithHealthCheck(() =>
      this.apiCall("/search-loras", HttpMethod.GET, params).then((response) =>
        response.json(),
      ),
    );
  };

  /**
   * Searches for audio models.
   * @param params - Object containing search query.
   * @returns Promise with search results.
   */
  searchAudioModels = async (params: { query: string }): Promise<AudioModel[]> => {
    return this.callWithHealthCheck(() =>
      this.apiCall("/search-voices-models", HttpMethod.GET, params).then(
        (response) => response.json(),
      ),
    );
  }

  /**
   * Generates an image based on parameters.
   * @param params - Object containing query and optional loraName.
   * @returns Promise with generation results.
   */
  generateImage = async (
    params: GenerationParams,
  ): Promise<GenerationResult> => {
    return this.callWithHealthCheck(() =>
      this.apiCall("/generateImage", HttpMethod.GET, params).then((response) =>
        response.json(),
      ),
    );
  };

  /**
   * Generates audio from text.
   * @param voiceModelId - The ID of the voice model to use.
   * @param text - The text to convert to speech.
   * @param pitch - The pitch of the voice (optional).
   * @returns Promise with generation results.
   */
  generateFromTTS = async (
    voiceModelName: string,
    text: string,
    pitch: number = 0,
  ): Promise<{result: string}> => {
    return this.callWithHealthCheck(() =>
      this.apiCall(
        "/voice",
        HttpMethod.POST,
        { voiceModelName, text, pitch: pitch.toString() },
      ).then((response) => response.json()),
    );
  };

  /**
   * Generates audio from an audio URL.
   * @param voiceModelId - The ID of the voice model to use.
   * @param audioUrl - The URL of the audio file to use as input.
   * @param pitch - The pitch of the voice (optional).
   * @returns Promise with generation results.
   */
  generateFromAudioURL = async (
    voiceModelName: string,
    audioUrl: string,
    pitch: number = 0,
  ): Promise<{result: string}> => {
    return this.callWithHealthCheck(() =>
      this.apiCall(
        "/voice",
        HttpMethod.POST,
        { voiceModelName, audioUrl, pitch: pitch.toString() },
      ).then((response) => response.json()),
    );
  };

  /**
   * Generates a progressive image based on parameters.
   * @param params - Object containing query and optional loraName.
   * @param callback - Function to call with status updates.
   * @returns Promise with generation results.
   */
  generateProgressiveImage = async (
    params: GenerationParams,
    callback: (status: ImageStatus, data: GenerationResult) => unknown = (
      status: ImageStatus,
    ) => {
      return status;
    },
  ) => {
    await this.getHealthData();

    const { imageId } = await this.generateImage(params);
    const statusResponse = await this.getStatus({ imageId });
    const { status } = statusResponse;
    callback(status, { imageId });
    if (status === ImageStatus.COMPLETED) {
      return statusResponse;
    }
    let oldModifiedDate: string | null = null;
    while (true) {
      await new Promise((resolve) => setTimeout(resolve, 100)); // Wait for 100 milliseconds
      const statusResponse = await this.getStatus({ imageId });
      const { status } = statusResponse;
      const lastModifiedDate = statusResponse.lastModifiedDate || null;
      const error = statusResponse.error || null;
      if (oldModifiedDate !== lastModifiedDate) {
        oldModifiedDate = lastModifiedDate;
        callback(status, { imageId });
      }

      if (status === ImageStatus.COMPLETED) {
        break;
      }

      if (status === ImageStatus.FAILED) {
        throw new Error("Image generation failed: " + error);
      }
    }
    return statusResponse;
  };
}
